cwe_checker_rs/src/analysis/graph.rs

```diff
@@ -242,14 +256,39 @@ impl<'a> GraphBuilder<'a> {
                             .add_edge(source, return_to_node, Edge::ExternCallStub(jump));
                     }
                 } else {
+                    let mut call_source_node: Option<NodeIndex> = None;
                     if let Some((target_node, _)) = self.call_targets.get(&target) {
-                        self.graph.add_edge(source, *target_node, Edge::Call(jump));
+                        let (target_block, target_sub) = match self.graph[*target_node] {
+                            Node::BlkStart(target_block, target_sub) => (target_block, target_sub),
+                            _ => panic!(),
+                        };
+                        call_source_node = Some(self.graph.add_node(Node::CallSource {
+                            source: (source_block, sub_term),
+                            target: (target_block, target_sub),
+                        }));
+                        self.graph.add_edge(
+                            source,
+                            *call_source_node.as_ref().unwrap(),
+                            Edge::CallCombine(jump),
+                        );
+                        self.graph.add_edge(
+                            *call_source_node.as_ref().unwrap(),
+                            *target_node,
+                            Edge::Call(jump),
+                        );
                     } // TODO: Log message for the else-case?
                     if let Some(return_node) = return_to_node_option {
-                        self.return_addresses
-                            .entry(target.clone())
-                            .and_modify(|vec| vec.push((source, return_node)))
-                            .or_insert_with(|| vec![(source, return_node)]);
+                        if let Some(cs_node) = call_source_node {
+                            self.return_addresses
+                                .entry(target.clone())
+                                .and_modify(|vec| vec.push((cs_node, return_node)))
+                                .or_insert_with(|| vec![(cs_node, return_node)]);
+                        } else {
+                            self.return_addresses
+                                .entry(target.clone())
+                                .and_modify(|vec| vec.push((source, return_node)))
+                                .or_insert_with(|| vec![(source, return_node)]);
+                        }
```

Comment:
I am pretty sure we do not want to do this. We use the `return_addresses` to add `CallReturn` nodes later on. Adding this case would mean we allow `CallReturn` nodes without corresponding `CallSource` nodes.  This would be an unwanted artifact in the control flow graph.